src-ext/TypeScript/ggvega/src/layerArrayByAes.ts

```diff
@@ -0,0 +1,164 @@
+ *  `layer` array Vega-Lite `layer` array, we have change the `encoding`
+ *  namespace to conform to Vega-Lite.
+ *
+ * **Called by**
+ *
+ * @see topLevelSpec
+ *
+ * **Calls**
+ *
+ * @see itmLayer
+ * @see itmLayerArrayByLabelObject
+ * @see itmLayerArrayByScaleArray
+ * @see itmLayerArrayByCoord
+ * @see layerByItmLayer
+ *
+ * @param ggDatasets `GG.ggDatasetsObject`, key-value pairs of ggspec datasets
+ * @param ggLayerArray `GG.Layer[]` - array of ggspec layers
+ * @param ggScaleArray `GG.Scale[]` - array of ggspec scales
+ * @param ggLabelObject `GG.Labels` - key-value pairs of ggspec labels
+ * @param ggCoordinates `GG.Coord` - ggspec coordinates
+ *
+ * @returns `vl.LayerSpec[]`, array containing Vega-Lite layer specs
+ *
+ */
+export function layerArrayByAes(
+  ggDatasetsObject: GG.DatasetsObject,
+  ggLayerArray: GG.Layer[],
+  ggScaleArray: GG.Scale[],
+  ggLabelObject: GG.Labels,
+  ggCoordinates: GG.Coord
+): VL.LayerSpec[] {
+  // validate
+  if (ggLayerArray.length == 0) {
+    throw new Error('ggplot object has no layers, requires at least one layer');
+  }
+
+  // translate
+
+  // start intermediate layers according to ggLayerArray
+  // could this work?
+  let itmLayerArray: ItmLayer[] = ggLayerArray.map((ggLayer: GG.Layer) => {
+    return itmLayer(ggLayer, ggDatasetsObject);
+  });
+
+  // incorporate labels
+  itmLayerArray = itmLayerArrayByLabelsObject(itmLayerArray, ggLabelObject);
+
+  // incorporate scales
+  itmLayerArray = itmLayerArrayByScalesArray(itmLayerArray, ggScaleArray);
+
+  // incorporate coordinates
+  itmLayerArray = itmLayerArrayByCoord(itmLayerArray, ggCoordinates);
+
+  // change encoding-key namespace from ggplot2 to Vega-Lite
+  const layerArray: VL.LayerSpec[] = itmLayerArray.map(layerByItmLayer);
+
+  return layerArray;
+}
+
+/**
+ * Change the `encoding` namespace
+ *
+ * @remark
+ * To translate an intermediate `layer` to a Vega-Lite `layer`, this function
+ * changes the `encoding` namespace from ggplo2 aesthetic-names to Vega-Lite
+ * `encoding` keys.
+ *
+ * As well, a ggspec `geom` is an element of an intermediate `layer`; we use it
+ * as a "breadcrumb", as the mapping of aesthetic names to encoding names depends
+ * on the `geom`.
+ *
+ * **Called by**
+ * @see layerArray
+ *
+ * **Calls**
+ * @see encodingNameByGeom
+ *
+ *
+ * @param itmLayer - `itmLayer`, intermediate layer-object
+ *
+ * @returns `VL.Layer`, Vega-Lite layer-object
+ *
+ */
+function layerByItmLayer(itmLayer: ItmLayer): VL.LayerSpec {
+  // create new encoding
+  var encoding: VL.Encoding = {};
+
+  // loop over aesthetic names in itmLayerEncoding
+  for (const aesName in itmLayer.encoding) {
+    if (Object.prototype.hasOwnProperty.call(itmLayer.encoding, aesName)) {
+      // get the encoding name, add to the encoding
+      const encodingName = encodingNameByGeom(aesName as GGEncodingKey, itmLayer.geomSet);
+      if (encodingName == 'x') encoding[encodingName] = itmLayer.encoding[aesName] as VL.XClass;
```

Comment:
I know I keep coming back to this code...

Could it be possible to create an Enum where the keys are the encoding names, e.g. `x` and the values are types, e.g.`VL.XClass`? Then could we cast to the type identified by the enum? 

This is more of a point to discuss than a call-to-action.

Reply:
I thought we had done this during the call on Tuesday 🤔